python/dgl/distributed/dist_graph.py

```diff
@@ -1075,26 +1075,53 @@ def _even_offset(n, k):
 def _split_even_to_part(partition_book, elements):
     ''' Split the input element list evenly.
     '''
+    # here we divide the element list as evenly as possible. If we use range partitioning,
+    # the split results also respect the data locality. Range partitioning is the default
+    # strategy.
+    # TODO(zhengda) we need another way to divide the list for other partitioning strategy.
     if isinstance(elements, DistTensor):
         # Here we need to fetch all elements from the kvstore server.
         # I hope it's OK.
         eles = F.nonzero_1d(elements[0:len(elements)])
+        # compute the offset of each split and ensure that the difference of each partition size
+        # is 1.
+        offsets = _even_offset(len(eles), partition_book.num_partitions())
+        assert offsets[-1] == len(eles)
+
+        # Get the elements that belong to the partition.
+        partid = partition_book.partid
+        part_eles = eles[offsets[partid] : offsets[partid + 1]]
     else:
-        eles = F.nonzero_1d(F.tensor(elements))
-
-    # here we divide the element list as evenly as possible. If we use range partitioning,
-    # the split results also respect the data locality. Range partitioning is the default
-    # strategy.
-    # TODO(zhengda) we need another way to divide the list for other partitioning strategy.
-
-    # compute the offset of each split and ensure that the difference of each partition size
-    # is 1.
-    offsets = _even_offset(len(eles), partition_book.num_partitions())
-    assert offsets[-1] == len(eles)
-
-    # Get the elements that belong to the partition.
-    partid = partition_book.partid
-    part_eles = eles[offsets[partid] : offsets[partid + 1]]
+        elements = F.tensor(elements)
```

Comment:
here `elements` is a vector stored in a local machine?
if we can store the entire `elements` array in the local machine, why do we still have memory issue?

Reply:
elements is a boolean mask to split train, validate and test sets (I learnt how to use it from this [example](https://github.com/dmlc/dgl/blob/master/examples/pytorch/graphsage/experimental/train_dist_unsupervised.py#L429-L430)). I agree this is not the best design. 
But in this case, elements is boolean tensor, so it cost 1/8 memory of the nonzero_1d tensor, so at least my machines have enough memory for it.
Do you have any plan about redesign the data split part to further reduce the memory cost?

Reply:
`elements` is a boolean vector so it consumes 8x less memory than an int64 tensor.  `nonzero` will give you an int64 tensor whose length is roughly the same order of magnitude.

See also https://discuss.dgl.ai/t/edge-split-consumes-too-much-memory/2130.

Reply:
i see. then my general comment is that we can unify the code for DistTensor and local tensor.

Reply:
Yes. But we need to implement a distributed count_nonzero method first. I can do it next week, what do you think?